src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreWorker.cs

```diff
@@ -208,9 +229,15 @@ private void SolutionEvents_AfterClosing()
             // on-board request onto pending restore operation
             _pendingRequests.Value.TryAdd(request);
 
+            // Await completion of the requested restore operation.
+            // The caller will be unblocked immediately upon
+            // cancellation request via provided token.
+            // Method returns false in case of cancellation is requested.
+            var cancellationTcs = new TaskCompletionSource<bool>();
+            using (token.Register(() => cancellationTcs.SetResult(false)))
```

Comment:
As inspired by [await anything](https://blogs.msdn.microsoft.com/pfxteam/2011/01/13/await-anything/). 

Didn't know about that particular extension method. Thanks for the tip!